# Assignment 1: TF.IDF & Cosine Similarity Programming Assignment

In [1]:
# Mary Abigail Garcia
# ID: 1085462
# CSCI4126 M02
# November 8, 2020

I will be answering the question of "How
many tokens and types(unique tokens) are in each document, and in the entire corpus(collections)?" in order from docs 1-10 and the entire corpus section by section. (Parsing document 1, 2, etc headings)

In [2]:
## Before starting the assignment, I want to combine all 10 text documents into one doc named "result.txt"

import glob

file_list = glob.glob("Data/*.txt")
with open('result.txt', 'w', encoding='utf8') as outfile:
    for filename in file_list:
        with open(filename, encoding='utf8') as infile:
            contents = infile.read()
            outfile.write(contents)

# Print the combined documents
combined_docs = open('result.txt', 'r', encoding='utf8')
for line in combined_docs:
    print(line.strip())

Gyarados


Gyarados (ギャラドス, Gyaradosu,  or ) is a Pokémon species in Nintendo and Game Freak's Pokémon franchise. Created by Ken Sugimori, Gyarados first appeared in the video games Pokémon Red and Pokemon Green and subsequent sequels, later appearing in various merchandise, spinoff titles and animated and printed adaptations of the franchise. Gyarados is voiced by Unshō Ishizuka in both Japanese and English media. Known as the Atrocious Pokémon, Gyarados is the evolved form of Magikarp and it is well known in the Pokémon world for its fierce temper as well as its reputation for causing nothing but destruction so much so that once it has worked itself into a frenzy, it will not calm down until everything around it has been destroyed.
Gyarados appears multiple times in the anime under various trainers such as Misty, Lance, Crasher Wake, and Nurse Joy. Two different Gyarados appear in the Pokémon Adventures manga. One is originally owned by Misty, but is traded between Red and Blue. The 

# Step 1: Tokenization

In [3]:
import nltk
import string

from collections import Counter

# define a function to tokenize the documents
# we remove the punctuations and normalize each word by making them lowercase
def get_tokens():
    with open(filename, encoding="utf8") as documents:
        text = documents.read()
        lowercase = text.lower()
        no_punctuation = text.translate(str.maketrans('','',string.punctuation))
        tokens = nltk.word_tokenize(no_punctuation)
    return tokens

#### Parsing document 1

In [4]:
filename = 'Data/a1.txt'
token1 = get_tokens()
count1 = Counter(token1)
len(token1)

1672

In [5]:
len(count1)

662

Number of tokens: 1672

Number of types: 662

#### Parsing document 2

In [6]:
filename = 'Data/a2.txt'
token2 = get_tokens()
count2 = Counter(token2)
len(token2)

4223

In [7]:
len(count2)

1338

Number of tokens: 4223

Number of types: 1338

#### Parsing document 3

In [8]:
filename = 'Data/a3.txt'
token3 = get_tokens()
count3 = Counter(token3)
len(token3)

1905

In [9]:
len(count3)

744

Number of tokens: 1905

Number of types: 744

#### Parsing document 4

In [10]:
filename = 'Data/a4.txt'
token4 = get_tokens()
count4 = Counter(token4)
len(token4)

2885

In [11]:
len(count4)

1033

Number of tokens: 2885

Number of types: 1033

#### Parsing document 5

In [12]:
filename = 'Data/a5.txt'
token5 = get_tokens()
count5 = Counter(token5)
len(token5)

2998

In [13]:
len(count5)

1113

Number of tokens: 2998

Number of types: 1113

#### Parsing document 6

In [14]:
filename = 'Data/a6.txt'
token6 = get_tokens()
count6 = Counter(token6)
len(token6)

1542

In [15]:
len(count6)

639

Number of tokens: 1542

Number of types: 639

#### Parsing document 7

In [16]:
filename = 'Data/a7.txt'
token7 = get_tokens()
count7 = Counter(token7)
len(token7)

1421

In [17]:
len(count7)

623

Number of tokens: 1421

Number of types: 623

#### Parsing document 8

In [18]:
filename = 'Data/a8.txt'
token8 = get_tokens()
count8 = Counter(token8)
len(token8)

1591

In [19]:
len(count8)

698

Number of tokens: 1591

Number of types: 698

#### Parsing document 9

In [20]:
filename = 'Data/a9.txt'
token9 = get_tokens()
count9 = Counter(token9)
len(token9)

1401

In [21]:
len(count9)

602

Number of tokens: 1401

Number of types: 602

#### Parsing document 10

In [22]:
filename = 'Data/a10.txt'
token10 = get_tokens()
count10 = Counter(token10)
len(token10)

1247

In [23]:
len(count10)

560

Number of tokens: 1247

Number of types: 560

#### Parsing the entire corpus(collections)

In [24]:
filename = 'result.txt'
token_all = get_tokens()
count_all = Counter(token_all)
len(token_all)

20885

In [25]:
len(count_all)

4144

Number of tokens: 20885

Number of types: 4144

# Step 2: Stopword Removal

How many tokens and types (unique tokens) are in each document after removing all the stop words?

#### Document 1

In [26]:
from nltk.corpus import stopwords

filtered1 = [w for w in token1 if not w in stopwords.words('english')]
count1 = Counter(filtered1)
len(filtered1)

1027

In [27]:
len(count1)

588

Number of tokens after stopword removal: 1027

Number of types after stopword removal: 588

#### Document 2

In [28]:
filtered2 = [w for w in token2 if not w in stopwords.words('english')]
count2 = Counter(filtered2)
len(filtered2)

2537

In [29]:
len(count2)

1236

Number of tokens after stopword removal: 2537

Number of types after stopword removal: 1236

#### Document 3

In [30]:
filtered3 = [w for w in token3 if not w in stopwords.words('english')]
count3 = Counter(filtered3)
len(filtered3)

1151

In [31]:
len(count3)

676

Number of tokens after stopword removal: 1151

Number of types after stopword removal: 676

#### Document 4

In [32]:
filtered4 = [w for w in token4 if not w in stopwords.words('english')]
count4 = Counter(filtered4)
len(filtered4)

1792

In [33]:
len(count4)

947

Number of tokens after stopword removal: 1792

Number of types after stopword removal: 947

#### Document 5

In [34]:
filtered5 = [w for w in token5 if not w in stopwords.words('english')]
count5 = Counter(filtered5)
len(filtered5)

1875

In [35]:
len(count5)

1031

Number of tokens after stopword removal: 1875

Number of types after stopword removal: 1031

#### Document 6

In [36]:
filtered6 = [w for w in token6 if not w in stopwords.words('english')]
count6 = Counter(filtered6)
len(filtered6)

941

In [37]:
len(count6)

565

Number of tokens after stopword removal: 941

Number of types after stopword removal: 565

#### Document 7

In [38]:
filtered7 = [w for w in token7 if not w in stopwords.words('english')]
count7 = Counter(filtered7)
len(filtered7)

886

In [39]:
len(count7)

561

Number of tokens after stopword removal: 886

Number of types after stopword removal: 561

#### Document 8

In [40]:
filtered8 = [w for w in token8 if not w in stopwords.words('english')]
count8 = Counter(filtered8)
len(filtered8)

989

In [41]:
len(count8)

624

Number of tokens after stopword removal: 989

Number of types after stopword removal: 624

#### Document 9

In [42]:
filtered9 = [w for w in token9 if not w in stopwords.words('english')]
count9 = Counter(filtered9)
len(filtered9)

867

In [43]:
len(count9)

527

Number of tokens after stopword removal: 867

 Number of types after stopword removal: 527

#### Document 10

In [44]:
filtered10 = [w for w in token10 if not w in stopwords.words('english')]
count10 = Counter(filtered10)
len(filtered10)

761

In [45]:
len(count10)

491

Number of tokens after stopword removal: 761

Number of types after stopword removal: 491

# Step 3: Stemming/Lemmatization

How many terms (size of vocabulary) are left in each document after stemming/lemmatization and what’s the total vocabulary size for the entire corpus?

In [46]:
from nltk.stem.porter import *

# Define the function to lemmatize the documents to its root words.

stemmer = PorterStemmer()
def stem_tokens(token, stemmer):
    stemmed=[]
    for item in token:
        stemmed.append(stemmer.stem(item))
    return stemmed

#### Document 1

In [47]:
stem_tokens(token1, stemmer)
stemmed = stem_tokens(filtered1, stemmer)
count1 = Counter(stemmed)
len(count1)

502

Size of Vocabulary: 502

#### Document 2

In [48]:
stem_tokens(token2, stemmer)
stemmed = stem_tokens(filtered2, stemmer)
count2 = Counter(stemmed)
len(count2)

997

Size of Vocabulary: 997

#### Document 3

In [49]:
stem_tokens(token3, stemmer)
stemmed = stem_tokens(filtered3, stemmer)
count3 = Counter(stemmed)
len(count3)

580

Size of Vocabulary: 580

#### Document 4

In [50]:
stem_tokens(token4, stemmer)
stemmed = stem_tokens(filtered4, stemmer)
count4 = Counter(stemmed)
len(count4)

775

Size of Vocabulary: 775

#### Document 5

In [51]:
stem_tokens(token5, stemmer)
stemmed = stem_tokens(filtered5, stemmer)
count5 = Counter(stemmed)
len(count5)

836

Size of Vocabulary: 836

#### Document 6

In [52]:
stem_tokens(token6, stemmer)
stemmed = stem_tokens(filtered6, stemmer)
count6 = Counter(stemmed)
len(count6)

498

Size of Vocabulary: 498

#### Document 7

In [53]:
stem_tokens(token7, stemmer)
stemmed = stem_tokens(filtered7, stemmer)
count7 = Counter(stemmed)
len(count7)

492

Size of Vocabulary: 492

#### Document 8

In [54]:
stem_tokens(token8, stemmer)
stemmed = stem_tokens(filtered8, stemmer)
count8 = Counter(stemmed)
len(count8)

528

Size of Vocabulary: 528

#### Document 9

In [55]:
stem_tokens(token9, stemmer)
stemmed = stem_tokens(filtered9, stemmer)
count9 = Counter(stemmed)
len(count9)

453

Size of Vocabulary: 453

#### Document 10

In [56]:
stem_tokens(token10, stemmer)
stemmed = stem_tokens(filtered10, stemmer)
count10 = Counter(stemmed)
len(count10)

427

Size of Vocabulary: 427

#### All the documents

In [57]:
# Perform stopword removal of documents as a whole
filtered_all = [w for w in token_all if not w in stopwords.words('english')]
count_all = Counter(filtered_all)

# Stemming/Lemmatization
stem_tokens(token_all, stemmer)
stemmed = stem_tokens(filtered_all, stemmer)
count_all = Counter(stemmed)
len(count_all)

2905

Size of Vocabulary: 2905

# Step 4: Compute TF-IDF of each document in the corpus

In [1]:
import nltk
import string
import os

from sklearn.feature_extraction.text import TfidfVectorizer
from nltk.stem.porter import PorterStemmer

path = 'Data'
token_dict = {}
stemmer = PorterStemmer()


# Apply lemmatization
def stem_tokens(tokens, stemmer):
    stemmed = []
    for item in tokens:
        stemmed.append(stemmer.stem(item))
    return stemmed

# Tokenize the text
def tokenize(text):
    tokens = nltk.word_tokenize(text)
    stems = stem_tokens(tokens, stemmer)
    return stems


for subdir, dirs, files in os.walk(path):
    for file in files:
        file_path = subdir + os.path.sep + file
        doc = open(file_path, 'r', encoding='utf8')
        text = doc.read()
        lowers = text.lower()
        no_punctuation = text.translate(str.maketrans('','',string.punctuation))
        token_dict[file] = no_punctuation

# I did not use stopwords parameter because it gave me warnings, so results may vary
tfidf = TfidfVectorizer(tokenizer=tokenize)
tfs = tfidf.fit_transform(token_dict.values())
print(tfs)

  (0, 380)	0.0039054358464855115
  (0, 1492)	0.0039054358464855115
  (0, 890)	0.0039054358464855115
  (0, 2119)	0.0039054358464855115
  (0, 2898)	0.010563220155141247
  (0, 697)	0.010563220155141247
  (0, 566)	0.006984705485077477
  (0, 1334)	0.010563220155141247
  (0, 1733)	0.008979702187436498
  (0, 251)	0.008979702187436498
  (0, 1774)	0.015712356618584142
  (0, 1450)	0.008979702187436498
  (0, 2230)	0.015712356618584142
  (0, 2938)	0.015712356618584142
  (0, 2660)	0.010563220155141247
  (0, 1034)	0.007856178309292071
  (0, 1541)	0.0042776636392283015
  (0, 856)	0.006984705485077477
  (0, 2533)	0.008979702187436498
  (0, 1038)	0.0046891423738825745
  (0, 2260)	0.006272660341587323
  (0, 1437)	0.008979702187436498
  (0, 2677)	0.006984705485077477
  (0, 2692)	0.010563220155141247
  (0, 2690)	0.008979702187436498
  :	:
  (9, 2446)	0.004751235131980083
  (9, 1622)	0.008675597505774813
  (9, 2772)	0.004337798752887406
  (9, 1446)	0.008675597505774813
  (9, 2301)	0.004751235131980083
  (9

# Step 5: Compute the Cosine Similarity

In [20]:
from sklearn.metrics.pairwise import cosine_similarity
import pandas

cos_sim = cosine_similarity(tfs, dense_output=False)
inOrder = pandas.DataFrame(cos_sim.toarray())
print('Computing Cosine Similarity for every pair of two documents in the corpus:\n {}\n'.format(inOrder))

Computing Cosine Similarity for every pair of two documents in the corpus:
           0         1         2         3         4         5         6  \
0  1.000000  0.521080  0.574955  0.525430  0.550008  0.509865  0.491978   
1  0.521080  1.000000  0.630418  0.568023  0.583114  0.553054  0.529054   
2  0.574955  0.630418  1.000000  0.638142  0.734219  0.662359  0.631943   
3  0.525430  0.568023  0.638142  1.000000  0.595270  0.571750  0.539223   
4  0.550008  0.583114  0.734219  0.595270  1.000000  0.608267  0.557317   
5  0.509865  0.553054  0.662359  0.571750  0.608267  1.000000  0.543021   
6  0.491978  0.529054  0.631943  0.539223  0.557317  0.543021  1.000000   
7  0.516628  0.551327  0.625272  0.561686  0.581911  0.554612  0.524662   
8  0.464698  0.496154  0.605459  0.516506  0.536971  0.517110  0.493548   
9  0.427660  0.455216  0.492442  0.471311  0.460873  0.443441  0.433049   

          7         8         9  
0  0.516628  0.464698  0.427660  
1  0.551327  0.496154  0.45521

In [21]:
with open("cossim.txt","w") as file:
    file.write(str(inOrder)) 
    file.close()

See cossim.txt for the similarity matrix of pairs of documents.